### DISCUSSION TITLE
Using sqlalchemy-firebird with firebird-driver and PyInstaller

### DISCUSSION THREAD

**Discussion by CatamountJack** [pauldex/sqlalchemy-firebird#67] [Q&A] [answered]
I fought with this issue a lot today and finally found a solution, so sharing the question and answer both here...

### Question:  
When trying to bundle my application that uses sqlalchemy-firebird and firebird-driver for the driver with PyInstaller, I get the following error when it tries to load:

`sqlalchemy.exc.NoSuchModuleError: Can't load plugin: sqlalchemy.dialects:firebird.firebird`

The program runs fine when not frozen, so it seems to be an issue with PyInstaller recognizing the drivers for firebird.

**Comment by CatamountJack** [comment, marked as answer]
### Answer

Be sure to manually register the dialect in your application.  I used the following code right before I created my connection string, but since the non-frozen application doesn't actually need it, it probably doesn't matter as much where it goes.  It just needs to be somewhere that PyInstaller will pick it up when it's freezing.

```python
from sqlalchemy.dialects import registry

registry.register(
        "firebird",
        "sqlalchemy_firebird.firebird",
        "FBDialect_firebird"
        )
registry.register(
        "firebird.firebird",
        "sqlalchemy_firebird.firebird",
        "FBDialect_firebird",
        )
```

I also included the following hidden-imports for PyInstaller (in my case, in my spec file).  I'm not sure if all of these are necessary, but I know at least `sqlalchemy_firebird.firebird` is.

```python
hidden_imports = [
        'sqlalchemy_firebird',
        'sqlalchemy_firebird.firebird',
        'firebird',
        'firebird.base',
        'firebird.driver',
        ]
```

This was using `SQLAlchemy 2.0.29`, `sqlalchemy-firebird 2.1`, `firebird-driver 1.10.1`, and `PyInstaller 6.5`

**Reply by mberdyshev** [reply]
Probably in your code there are no imports of `sqlalchemy_firebird` or `firebird_driver`, as it is usually used just as a plugin. In this case, `PyInstaller` can not spot that these libraries should be included in the bundle. This is where hidden imports are helpful.
You can read more about it [here](https://pyinstaller.org/en/stable/when-things-go-wrong.html#listing-hidden-imports).
